services/device-connection/src/main/java/org/eclipse/hono/deviceconnection/infinispan/RemoteCacheBasedDeviceConnectionService.java

```diff
@@ -0,0 +1,217 @@
+
+    private static final Logger LOG = LoggerFactory.getLogger(RemoteCacheBasedDeviceConnectionService.class);
+    private RemoteCacheContainer cacheManager;
+    private BasicCache<String, String> cache;
+
+    /**
+     * Sets the cache manager to use for retrieving a cache.
+     * 
+     * @param cacheManager The cache manager.
+     * @throws NullPointerException if cache manager is {@code null}.
+     */
+    @Autowired
+    public void setCacheManager(final RemoteCacheContainer cacheManager) {
+        this.cacheManager = Objects.requireNonNull(cacheManager);
+        log.info("using cache manager [{}]", cacheManager.getClass().getName());
+    }
+
+    void setCache(final BasicCache<String, String> cache) {
+        this.cache = cache;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void doStart(final Future<Void> startFuture) {
+
+        if (cacheManager == null) {
+            startFuture.fail(new IllegalStateException("cache manager is not set"));
+        } else {
+            context.executeBlocking(result -> {
+                cacheManager.start();
+                cache = cacheManager.getCache("device-connection");
+                cache.start();
+                result.complete(cacheManager);
+            }, attempt -> {
+                if (attempt.succeeded()) {
+                    LOG.info("successfully connected to remote cache");
+                } else {
+                    LOG.info("failed to connect to remote cache", attempt.cause());
+                }
+            });
+            startFuture.complete();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void doStop(final Future<Void> stopFuture) {
+        context.executeBlocking(result -> {
+            if (cacheManager != null) {
+                cacheManager.stop();
+            }
+            result.complete();
+        }, stopAttempt -> {
+            if (stopAttempt.succeeded()) {
+                LOG.info("connection(s) to remote cache stopped successfully");
+            } else {
+                LOG.info("error trying to stop connection(s) to remote cache", stopAttempt.cause());
+            }
+            stopFuture.complete();
+        });
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void setLastKnownGatewayForDevice(
+            final String tenantId,
+            final String deviceId,
+            final String gatewayId,
+            final Span span,
+            final Handler<AsyncResult<DeviceConnectionResult>> resultHandler) {
+
+        if (cache == null) {
+            resultHandler.handle(Future.failedFuture(new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "no connection to remote cache")));
+        } else {
+            cache
+            .putAsync(getKey(tenantId, deviceId), gatewayId)
```

Comment:
I think it would make sense to set a TTL here. And allow for automatic cache cleanup.

Reply:
@calohmn and I discussed using an expiring cache as well but in the end thought that the data itself will be valid (and useful) until being updated. Based on that, the data does not seem to be of an ephemeral nature really. This is also one of the reasons for using a distributed/sharded cache which can scale out horizontally, i.e. without the need to keep all data on each node of the cluster (fully replicated).
However, what would be a reasonable value for a TTL from your point of view?

Reply:
I think that you need some way to clean up the cache eventually. The entry is no longer valid, when the device is no longer registered with the system. Whatever the reason for that is.

You can create some elaborate cleanup mechanism, or simply set a TTL of e.g. 1 month.

Reply:
I agree that the entry becomes irrelevant once the device is removed from the system and I do agree that implementing a sophisticated mechanism for removing entries for deleted devices would be a little over the top for now. So let's start with a TTL of 1 month and see how we fare ...

Reply:
I have pushed a corresponding change ...